jax/linear_util.py

```diff
@@ -12,6 +12,59 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+"""
+Utilities for defining linear functions composed with transformations.
+
+"Linear" here is meant in the sense of linear types; that is, a linear function
+may be called at most once.
+
+For example:
+
+from jax import linear_util as lu
+
+# A transformation that scales its argument down and its result up.
+@lu.transformation
+def scale_transformer(scale, x):
+  ans = yield (x / scale,)
+  yield (x * scale,)
+
+def f(x):
+  return x + 1
+
+g = lu.wrap_init(f)  # Wraps `f` as a `WrappedFun`.
+g = scale_transformer(g, 2.0)  # Scale inputs/outputs by 2.0
+g = scale_transformer(g, 0.7)  # Scale inputs/outputs further by 0.7.
+print(g.call_wrapped(3.))  # Call the transformed function.
+
+
+A `WrappedFun` object represents a function `f`, together with a
+sequence of nested transformations that to be applied to the positional
+arguments at call time and function return values at return time.
+`WrappedFun` objects explicitly represent the set of transformations so that
+they can be used as dictionary keys for memoization. `WrappedFun` objects
+compare as equal only if they compute the same function.
+
+Transformations are implemented as generators to save call stack frames.
+A transformation's generator takes arguments `gen args + args`, and yields
+a tuple of transformed arguments that should be passed to the wrapped
+function. The result of the wrapped function is passed back to the generator
+using gen.send(), and the generator yields the transformed results to pass
+back to the caller.
+
+Transformations can also return auxiliary data using the `transform_with_aux`
+decorator. For example:
+
+@lu.transformation_with_aux
+def scale_transformer_aux(scale, x):
+  ans = yield (x / scale,)
+  yield ((x * scale,), "Auxiliary data: {}".format(x))
```

Comment:
Same here, I think you can return `x * scale, "Aux data"` though correct me if I'm wrong.

Reply:
Done.